### PR TITLE
Only allow using the atomic intrinsics on integer types

### PR DESCRIPTION
**Pull request by Amanieu**
Using these with non-integer types results in LLVM asserts. Atomic operations on non-integer types will require values be transmuted into an integer type of suitable size.

This doesn't affect the standard library since `AtomicBool` and `AtomicPtr` currently use `usize` for atomic operations.

r? @eddyb

### CODE REVIEW

**Comment by eddyb** [comment]
Can you add a compile-fail test for the errors?

**Comment by Amanieu** [comment]
Done (and fixed a few bugs)

**Comment by Amanieu** [comment]
Updated to fix make check complaints about line length

**Comment by Amanieu** [comment]
Fixed more line lengths...

**Comment by Amanieu** [comment]
Fixed the test

**Comment by eddyb** [comment]
@Amanieu ah, you're hitting the issue where MIR lacked precise spans. Now they're there, but not used.
If you want to fix it, have `trans_intrinsic_call` take a `span: Span` argument instead of extracting it from `call_debug_location`, and pass `data.terminator()` for that argument, from `MirContext::trans_block`.
You can then also remove `#[rustc_no_mir]` from `src/test/compile-fail/simd-intrinsic-*.rs`.
Alternatively, you can just do what I did in those tests and we'll fix this when we implement debuginfo for MIR.

**Comment by Amanieu** [comment]
@eddyb I went with the easy option for now...

File: src/test/compile-fail/non-interger-atomic.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#![allow(warnings)]
+
+use std::intrinsics;
+
+#[derive(Copy, Clone)]
+struct Foo(i64);
+type Bar = &'static Fn();
+type Quux = [u8; 100];
+
+#[rustc_no_mir] // FIXME #27840 MIR doesn't provide precise spans for calls.
```
Comment (eddyb) [inline src/test/compile-fail/non-interger-atomic.rs]: You need `#[feature(rustc_attrs)]` to use this attribute.

Comment (Amanieu) [inline src/test/compile-fail/non-interger-atomic.rs]: Oops, fixed.

**Comment by eddyb** [comment]
@bors force
Looks like a timeout. Doing this to avoid building everything again.

**Pull request merged**